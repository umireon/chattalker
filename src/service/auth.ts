import type { AppContext } from '../../constants'
import type { Auth } from 'firebase/auth'
import { signInWithCustomToken } from 'firebase/auth'

interface AuthenticateWithTokenOptions {
  readonly token: string
  readonly uid: string
}

export const authenticateWithToken = async (auth: Auth, { authenticateWithTokenEndpoint }: AppContext, { token, uid }: AuthenticateWithTokenOptions) => {
  const query = new URLSearchParams({ token, uid })
  const response = await fetch(`${authenticateWithTokenEndpoint}?${query}`)
  if (!response.ok) throw new Error('Authentication failed')
  const customToken = await response.text()
  const credential = await signInWithCustomToken(auth, customToken)
  return credential
}
