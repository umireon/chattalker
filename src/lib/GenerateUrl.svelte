<script lang="ts">
  import { getUserData, setUserData } from '../service/users'

  import type { Firestore } from 'firebase/firestore'
  import type { User } from 'firebase/auth'

  export let db: Firestore
  export let user: User

  let value: string = ''

  function uint8ArrayToHexString (array: Uint8Array) {
    return Array.from(array, b => b.toString(16).padStart(2, '0')).join('')
  }

  async function handleClickGenerateUrl () {
    const { token } = await getUserData(db, user)
    if (typeof token !== 'undefined' && token !== null) {
      const query = new URLSearchParams({ token: token, uid: user.uid })
      value = `${location.origin}${location.pathname}#${query}`
    } else {
      const buffer = new Uint8Array(256)
      const newTokenArray = crypto.getRandomValues(buffer)
      const newToken = uint8ArrayToHexString(newTokenArray)
      await setUserData(db, user, { token: newToken })
      const query = new URLSearchParams({ token: newToken, uid: user.uid })
      value = `${location.origin}${location.pathname}#${query}`
    }
  }

  async function handleClickCopyUrl () {
    await navigator.clipboard.writeText(value)
  }


  async function handleClickResetUrl () {
    await setUserData(db, user, { token: null })
    value = ''
  }
</script>

<main>
  <p>
    <button type="button" on:click={handleClickGenerateUrl}>Generate URL for OBS browser</button>
    <input placeholder="OBS URL" {value}>
    <button type="button" on:click={handleClickCopyUrl}>Copy URL</button>
    <button type="button" on:click={handleClickResetUrl}>Reset URL</button>
  </p>
</main>
