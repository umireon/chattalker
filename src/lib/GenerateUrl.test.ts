import { expect, test } from 'vitest'

import { type Firestore } from "firebase/firestore";
import GenerateUrl from "./GenerateUrl.svelte";
import { type User } from "firebase/auth";
import { render } from "@testing-library/svelte";

test("GenerateUrl snapshot", () => {
  const props = { db: {} as Firestore, user: {} as User };
  const component = render(GenerateUrl, { props });
  expect(component.container).toMatchSnapshot();
});
