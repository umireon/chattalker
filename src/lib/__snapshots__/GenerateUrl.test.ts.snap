// Vitest Snapshot v1

exports[`GenerateUrl snapshot 1`] = `
<body>
  <div>
    <main>
      <p>
        <button
          type="button"
        >
          Generate URL for OBS browser
        </button>
         
        <input
          placeholder="OBS URL"
        />
         
        <button
          type="button"
        >
          Copy URL
        </button>
         
        <button
          type="button"
        >
          Reset URL
        </button>
      </p>
    </main>
  </div>
</body>
`;
